File: apkEditor/src/main/java/com/mcal/apkeditor/ApkComposeService.java
package com.mcal.apkeditor;

import android.annotation.SuppressLint;
import android.app.NotificationManager;
import android.app.PendingIntent;
import android.app.Service;
import android.content.Context;
import android.content.DialogInterface;
import android.content.Intent;
import android.graphics.Bitmap;
import android.graphics.BitmapFactory;
import android.os.Binder;
import android.os.Handler;
import android.os.IBinder;
import android.os.Message;
import android.util.Log;

import androidx.annotation.NonNull;
import androidx.appcompat.app.AlertDialog;
import androidx.core.app.NotificationCompat;

import com.google.android.material.dialog.MaterialAlertDialogBuilder;
import com.mcal.apkeditor.activities.ApkComposeActivity;
import com.mcal.apkeditor.ce.IApkMaking;
import com.mcal.apkeditor.data.Constants;
import com.mcal.apkeditor.utils.AssetsInstaller;
import com.mcal.common.data.Preferences;
import com.mcal.common.utils.ActivityUtils;
import com.mcal.common.utils.ITaskCallback;
import com.mcal.seticon.SetIcon;

import java.io.BufferedReader;
import java.io.File;
import java.io.FileInputStream;
import java.io.IOException;
import java.io.InputStreamReader;
import java.lang.ref.WeakReference;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.HashSet;
import java.util.List;
import java.util.Map;
import java.util.Set;

import brut.androlib.Androlib;
import brut.androlib.options.BuildOptions;

public class ApkComposeService extends Service implements ITaskCallback {
    private final ComposeResult composeResult = new ComposeResult();
    // For UI management
    private final MyHandler handler = new MyHandler();
    private final ComposeServiceBinder binder = new ComposeServiceBinder();
    private String decodeRootPath;
    private String srcApkPath;
    private Boolean stringModified;
    private Boolean manifestModified;
    private Boolean resFileModified;
    private ArrayList<String> modifiedSmaliFolders;
    // private ArrayList<String> replaceEntries;
    // private ArrayList<String> replaceFiles;
    private Map<String, String> addedFiles;
    private Map<String, String> replacedFiles;
    private Set<String> deletedFiles;
    // Recorded all the relation between file entry to zip entry
    // like res/drawable-hdpi-v4/a.png -> res/drawable-hdpi/a.png
    private Map<String, String> fileEntry2ZipEntry;
    // Output apk path
    private String targetApkPath;
    // Composing thread and result
    private ComposeThread composeThread;
    private WeakReference<ITaskCallback> observer;
    private IApkMaking extraMaker = null;
    // Foreground notification
    private NotificationManager mNotificationManager;
    private NotificationCompat.Builder mNotifyBuilder;
    private boolean foregroundStarted = false;
    private boolean signAPK;
    // Update notification title and description
    // Called from non-UI thread
    private long lastUpdateTime = 0;

    @Override
    public void onCreate() {
        super.onCreate();
    }

    @Override
    public int onStartCommand(Intent intent, int flags, int startId) {
        // When the service is restarted, intent = null
        if (intent == null) {
            return super.onStartCommand(intent, flags, startId);
        }

        this.decodeRootPath = ActivityUtils.getParam(intent, "decodeRootPath");
        this.srcApkPath = ActivityUtils.getParam(intent, "srcApkPath");
        this.targetApkPath = ActivityUtils.getParam(intent, "targetApkPath");
        String str = ActivityUtils.getParam(intent, "stringModified");
        this.stringModified = Boolean.valueOf(str);
        str = ActivityUtils.getParam(intent, "manifestModified");
        this.manifestModified = Boolean.valueOf(str);
        str = ActivityUtils.getParam(intent, "resFileModified");
        this.resFileModified = Boolean.valueOf(str);
        this.modifiedSmaliFolders = ActivityUtils.getStringArray(intent, "modifiedSmaliFolders");
        this.signAPK = ActivityUtils.getBoolParam(intent, "signAPK");

        this.addedFiles = ActivityUtils.getMapParam(intent, "addedFiles");
        this.replacedFiles = ActivityUtils.getMapParam(intent, "replacedFiles");
        this.deletedFiles = new HashSet<>();
        List<String> delEntries = ActivityUtils.getStringArray(intent, "deletedFiles");
        this.deletedFiles.addAll(delEntries);

        String passedFile = ActivityUtils.getParam(intent, "fileEntry2ZipEntry");
        if (passedFile != null) {
            this.fileEntry2ZipEntry = getMapFromFile(passedFile);
        }

        resetStatus();

        // Initially show notification in pro version
        // For free version, only show it when ad is ready
        if (BuildConfig.IS_PRO) {
            showNotification();
        }

        startComposeThread();

        return START_STICKY;
    }

    @Override
    public void onDestroy() {
        super.onDestroy();
    }

    private void showNotification() {
        Intent composeIntent = new Intent(this, ApkComposeActivity.class);
        composeIntent.setAction(Constants.ACTION.MAIN_ACTION);
        composeIntent.setFlags(Intent.FLAG_ACTIVITY_REORDER_TO_FRONT);
        PendingIntent pendingIntent = PendingIntent.getActivity(this, 0, composeIntent, 0);

        int iconId = SetIcon.getIconId();
        Bitmap icon = BitmapFactory.decodeResource(getResources(), iconId);
        String appName = getString(R.string.app_name);

        this.mNotificationManager =
                (NotificationManager) getSystemService(Context.NOTIFICATION_SERVICE);

        this.mNotifyBuilder = new NotificationCompat.Builder(this, ApkComposeActivity.PRIMARY_NOTIF_CHANNEL);
        this.mNotifyBuilder.setContentTitle(appName)
                .setTicker(appName)
                .setContentText(getString(R.string.build_ongoing))
                .setSmallIcon(iconId)
                .setLargeIcon(Bitmap.createScaledBitmap(icon, 128, 128, false))
                .setContentIntent(pendingIntent)
                .setOngoing(true);
        startForeground(Constants.NOTIFICATION_ID.FOREGROUND_SERVICE, mNotifyBuilder.build());
        this.foregroundStarted = true;
    }

    private void updateNotification(boolean forceShow, String title, String desc) {
        if (mNotificationManager != null) {
            handler.removeMessages(0);
            handler.setInfo(title, desc);
            // If not updated for a long time, then directly update it
            if (forceShow || (System.currentTimeMillis() - lastUpdateTime > 1000)) {
                handler.sendEmptyMessage(0);
            }
            // Otherwise send delayed message
            else {
                handler.sendEmptyMessageDelayed(0, 500);
            }
        }
    }

    private void resetStatus() {
        if (composeThread != null && composeThread.isAlive()) {
            composeThread.stopRunning();
        }

        composeResult.clear();
    }

    private void startComposeThread() {
//        if(Preferences.isApkToolCompiler()) {
//            MaterialAlertDialogBuilder dialog = new MaterialAlertDialogBuilder(this);
//            dialog.setTitle("Build Mode");
//            //dialog.setView("");
//            dialog.setPositiveButton("Build", (dialogInterface, i) -> startBuild());
//            dialog.setNegativeButton(android.R.string.cancel, null);
//        } else {
            startBuild();
//        }
    }

    public void startBuild() {
            if (srcApkPath != null) {
                this.composeThread = new ApkComposeThread(this, decodeRootPath,
                        srcApkPath, targetApkPath);
            } else {
                // srcApkPath == null, means currently is a full decoding
                this.composeThread = new ApkComposeThreadNew(this, decodeRootPath, targetApkPath);
            }

            if (extraMaker != null) {
                composeThread.setExtraMaker(extraMaker);
            }
            composeThread.setModification(this.stringModified,
                    this.manifestModified, this.resFileModified,
                    this.modifiedSmaliFolders, this.addedFiles, this.replacedFiles,
                    this.deletedFiles, this.fileEntry2ZipEntry, this.signAPK);
            composeThread.setTaskCallback(this);
            composeThread.start();
    }

    @NonNull
    private Map<String, String> getMapFromFile(String filepath) {
        Map<String, String> result = new HashMap<>();
        BufferedReader br = null;
        try {
            br = new BufferedReader(
                    new InputStreamReader(new FileInputStream(filepath)));
            String line = br.readLine();
            while (line != null) {
                String key = line;
                String value = br.readLine();
                if (value != null) {
                    result.put(key, value);
                } else {
                    break;
                }

                line = br.readLine();
            }
        } catch (Exception e) {
            e.printStackTrace();
        } finally {
            if (br != null) {
                try {
                    br.close();
                } catch (IOException e) {
                    e.printStackTrace();
                }
            }
        }
        return result;
    }

    @Override
    public IBinder onBind(Intent intent) {
        return binder;
    }

    @Override
    public void setTaskStepInfo(TaskStepInfo stepInfo) {
        synchronized (composeResult) {
            this.composeResult.curStep = stepInfo;
        }

        if (observer != null) {
            ITaskCallback ob = observer.get();
            if (ob != null) {
                ob.setTaskStepInfo(stepInfo);
            }
        }

        String desc = String.format(
                getResources().getString(R.string.step) + " %d/%d: %s",
                stepInfo.stepIndex, stepInfo.stepTotal,
                stepInfo.stepDescription);

        boolean forceShow = (stepInfo.stepIndex == stepInfo.stepTotal);
        updateNotification(forceShow, getString(R.string.build_ongoing), desc);
    }

    @Override
    public void setTaskProgress(float progress) {
        if (observer != null) {
            ITaskCallback ob = observer.get();
            if (ob != null) {
                ob.setTaskProgress(progress);
            }
        }
    }

    @Override
    public void taskSucceed() {
        synchronized (composeResult) {
            this.composeResult.finished = true;
            this.composeResult.ret = true;
            this.composeResult.failMessage = null;
        }

        updateNotification(true, getString(R.string.build_finished), getString(R.string.succeed));

        if (observer != null) {
            ITaskCallback ob = observer.get();
            if (ob != null) {
                ob.taskSucceed();
            }
        }
    }

    @Override
    public void taskFailed(String errMessage) {
        synchronized (composeResult) {
            this.composeResult.finished = true;
            this.composeResult.ret = false;
            this.composeResult.failMessage = errMessage;
        }

        updateNotification(true, getString(R.string.build_finished), getString(R.string.failed));

        if (observer != null) {
            ITaskCallback ob = observer.get();
            if (ob != null) {
                ob.taskFailed(errMessage);
            }
        }
    }

    @Override
    public void taskWarning(String message) {
        if (observer != null) {
            ITaskCallback ob = observer.get();
            if (ob != null) {
                ob.taskWarning(message);
            }
        }
    }

    public static class ComposeResult {
        public boolean finished = false;
        public boolean ret;
        public String failMessage;
        public TaskStepInfo curStep;

        public void clear() {
            finished = false;
            failMessage = null;
            curStep = null;
        }
    }

    @SuppressLint("HandlerLeak")
    private class MyHandler extends Handler {
        private String title;
        private String desc;

        public void setInfo(String title, String desc) {
            this.title = title;
            this.desc = desc;
        }

        @Override
        public void handleMessage(@NonNull Message msg) {
            switch (msg.what) {
                case 0:
                    if (title != null) {
                        mNotifyBuilder.setContentTitle(title);
                    }
                    if (desc != null) {
                        mNotifyBuilder.setContentText(desc);
                    }

                    if (foregroundStarted) {
                        mNotificationManager.notify(
                                Constants.NOTIFICATION_ID.FOREGROUND_SERVICE,
                                mNotifyBuilder.build());
                    } else {
                        startForeground(
                                Constants.NOTIFICATION_ID.FOREGROUND_SERVICE,
                                mNotifyBuilder.build());
                        foregroundStarted = true;
                    }
                    lastUpdateTime = System.currentTimeMillis();
                    break;
            }
        }
    }

    public class ComposeServiceBinder extends Binder {
        public void setObserver(ITaskCallback _observer) {
            observer = new WeakReference<>(_observer);
            synchronized (composeResult) {
                if (composeResult.finished) {
                    if (composeResult.ret) {
                        _observer.taskSucceed();
                    } else {
                        _observer.taskFailed(composeResult.failMessage);
                    }
                } else {
                    if (composeResult.curStep != null) {
                        _observer.setTaskStepInfo(composeResult.curStep);
                    }
                }
            }
        }

        public void buildAgain() {
            resetStatus();
            showNotification();
            startComposeThread();
        }

        // Stop the build thread
        public void stopBuilding() {
            if (composeThread != null && composeThread.isAlive()) {
                composeThread.stopRunning();
            }
            hideNotification();
        }

        public void setBuildHooker(IApkMaking extraMaker) {
            ApkComposeService.this.extraMaker = extraMaker;
        }

        // Get key/value maps
        public Map<String, Object> getValues() {
            Map<String, Object> ret = new HashMap<>();
            ret.put("srcApkPath", srcApkPath);
            ret.put("targetApkPath", targetApkPath);
            ret.put("decodeRootPath", decodeRootPath);
            boolean codeModified = (modifiedSmaliFolders != null && !modifiedSmaliFolders.isEmpty());
            ret.put("codeModified", codeModified);
            ret.put("signAPK", signAPK);
            return ret;
        }

        public void hideNotification() {
            if (mNotificationManager != null) {
                mNotificationManager.cancel(Constants.NOTIFICATION_ID.FOREGROUND_SERVICE);
                if (foregroundStarted) {
                    stopForeground(true);
                    foregroundStarted = false;
                }
                mNotificationManager = null;
                Log.e("DEBUG", "notification hided.");
            }
        }

        public void showNotification() {
            ApkComposeService.this.showNotification();
        }

        // Build thread is still running
        public boolean isRunning() {
            return (composeThread != null && composeThread.isAlive());
        }
    }
}

File: apkEditor/src/main/java/com/mcal/apkeditor/ApkComposeService.kt
package com.mcal.apkeditor

import android.annotation.SuppressLint
import android.app.NotificationManager
import android.app.PendingIntent
import android.app.Service
import android.content.Intent
import android.os.*
import android.util.Log
import androidx.core.app.NotificationCompat
import com.mcal.apkeditor.activities.ApkComposeActivity
import com.mcal.apkeditor.ce.IApkMaking
import com.mcal.apkeditor.data.Constants
import com.mcal.common.utils.ActivityHelper
import com.mcal.common.utilsOld.ITaskCallback
import com.mcal.common.utilsOld.ITaskCallback.TaskStepInfo
import kotlinx.coroutines.isActive
import java.io.BufferedReader
import java.io.FileInputStream
import java.io.IOException
import java.io.InputStreamReader
import java.lang.ref.WeakReference

class ApkComposeService : Service(), ITaskCallback {
    private val composeResult = ComposeResult()

    // For UI management
    private val handler: MyHandler = MyHandler()
    private val binder = ComposeServiceBinder()
    private var decodeRootPath: String? = null
    private var srcApkPath: String? = null
    private var stringModified: Boolean? = null
    private var manifestModified: Boolean? = null
    private var resFileModified: Boolean? = null
    private var modifiedSmaliFolders: ArrayList<String>? = null

    private var addedFiles: MutableMap<String, String>? = null
    private var replacedFiles: MutableMap<String, String>? = null
    private var deletedFiles: MutableSet<String>? = null

    // Recorded all the relation between file entry to zip entry
    // like res/drawable-hdpi-v4/a.png -> res/drawable-hdpi/a.png
    private var fileEntry2ZipEntry: Map<String, String>? = null

    // Output apk path
    private var targetApkPath: String? = null

    // Composing thread and result
    private var composeThread: ComposeThread? = null
    private var observer: WeakReference<ITaskCallback>? = null
    private var extraMaker: IApkMaking? = null

    // Foreground notification
    private var mNotificationManager: NotificationManager? = null
    private var mNotifyBuilder: NotificationCompat.Builder? = null
    private var foregroundStarted = false
    private var signAPK = false

    // Update notification title and description
    // Called from non-UI thread
    private var lastUpdateTime: Long = 0

    override fun onStartCommand(intent: Intent, flags: Int, startId: Int): Int {
        // When the service is restarted, intent = null
        decodeRootPath = ActivityHelper.getParam(intent, "decodeRootPath")
        srcApkPath = ActivityHelper.getParam(intent, "srcApkPath")
        targetApkPath = ActivityHelper.getParam(intent, "targetApkPath")
        var str = ActivityHelper.getParam(intent, "stringModified")
        stringModified = str.toBoolean()
        str = ActivityHelper.getParam(intent, "manifestModified")
        manifestModified = str.toBoolean()
        str = ActivityHelper.getParam(intent, "resFileModified")
        resFileModified = str.toBoolean()
        modifiedSmaliFolders = ActivityHelper.getStringArray(intent, "modifiedSmaliFolders")
        signAPK = ActivityHelper.getBoolParam(intent, "signAPK")
        addedFiles = ActivityHelper.getMapParam(intent, "addedFiles")
        replacedFiles = ActivityHelper.getMapParam(intent, "replacedFiles")
        deletedFiles = HashSet()
        val delEntries: List<String>? = ActivityHelper.getStringArray(intent, "deletedFiles")
        delEntries?.let { entries ->
            deletedFiles?.addAll(entries)
        }
        val passedFile = ActivityHelper.getParam(intent, "fileEntry2ZipEntry")
        if (passedFile != null) {
            fileEntry2ZipEntry = getMapFromFile(passedFile)
        }
        resetStatus()

        // Initially show notification in pro version
        // For free version, only show it when ad is ready
        if (BuildConfig.IS_PRO) {
            showNotification()
        }
        startComposeThread()
        return START_STICKY
    }

    private fun showNotification() {
        val composeIntent = Intent(this, ApkComposeActivity::class.java)
        composeIntent.action = Constants.ACTION.MAIN_ACTION
        composeIntent.flags = Intent.FLAG_ACTIVITY_REORDER_TO_FRONT
        val pendingIntent = PendingIntent.getActivity(this, 0, composeIntent, PendingIntent.FLAG_IMMUTABLE or PendingIntent.FLAG_UPDATE_CURRENT)
        val iconId = R.mipmap.ic_launcher_round
        val appName = getString(R.string.app_name)
        mNotificationManager = getSystemService(NOTIFICATION_SERVICE) as NotificationManager
        mNotifyBuilder = NotificationCompat.Builder(this, ApkComposeActivity.PRIMARY_NOTIF_CHANNEL)
        mNotifyBuilder?.let { builder ->
            builder.setContentTitle(appName)
                .setTicker(appName)
                .setContentText(getString(R.string.build_ongoing))
                .setSmallIcon(iconId)
                .setContentIntent(pendingIntent)
                .setOngoing(true)
            startForeground(Constants.NOTIFICATION_ID.FOREGROUND_SERVICE, builder.build())
        }
        foregroundStarted = true
    }

    private fun updateNotification(forceShow: Boolean, title: String, desc: String) {
        if (mNotificationManager != null) {
            handler.removeMessages(0)
            handler.setInfo(title, desc)
            // If not updated for a long time, then directly update it
            if (forceShow || System.currentTimeMillis() - lastUpdateTime > 1000) {
                handler.sendEmptyMessage(0)
            } else {
                handler.sendEmptyMessageDelayed(0, 500)
            }
        }
    }

    private fun resetStatus() {
        composeThread?.let { thread ->
            if (thread.isActive) {
                thread.stopRunning()
            }
        }
        composeResult.clear()
    }

    private fun startComposeThread() {
        composeThread = if (srcApkPath != null) {
            ApkComposeThread(
                this, decodeRootPath,
                srcApkPath, targetApkPath
            )
        } else {
            // srcApkPath == null, means currently is a full decoding
            decodeRootPath?.let { decodePath ->
                targetApkPath?.let { apkPath ->
                    ApkComposeThreadNew(this, decodePath, apkPath)
                }
            }
        }
        composeThread?.let { thread ->
            if (extraMaker != null) {
                thread.setExtraMaker(extraMaker)
            }

            stringModified?.let { string ->
                manifestModified?.let { manifest ->
                    resFileModified?.let { res ->
                        thread.setModification(
                            string, manifest, res,
                            modifiedSmaliFolders, addedFiles, replacedFiles,
                            deletedFiles, fileEntry2ZipEntry, signAPK
                        )
                    }
                }
            }
            thread.setTaskCallback(this)
            thread.execute()
        }
    }

    private fun getMapFromFile(filepath: String): Map<String, String> {
        val result: MutableMap<String, String> = HashMap()
        var br: BufferedReader? = null
        try {
            br = BufferedReader(
                InputStreamReader(FileInputStream(filepath))
            )
            var line = br.readLine()
            while (line != null) {
                val key = line
                val value = br.readLine()
                if (value != null) {
                    result[key] = value
                } else {
                    break
                }
                line = br.readLine()
            }
        } catch (e: Exception) {
            e.printStackTrace()
        } finally {
            if (br != null) {
                try {
                    br.close()
                } catch (e: IOException) {
                    e.printStackTrace()
                }
            }
        }
        return result
    }

    override fun onBind(intent: Intent): IBinder {
        return binder
    }

    override fun setTaskStepInfo(stepInfo: TaskStepInfo) {
        synchronized(composeResult) { composeResult.curStep = stepInfo }
        observer?.get()?.setTaskStepInfo(stepInfo)
        val desc = String.format(
            resources.getString(R.string.step) + " %d/%d: %s",
            stepInfo.stepIndex, stepInfo.stepTotal,
            stepInfo.stepDescription
        )
        val forceShow = stepInfo.stepIndex == stepInfo.stepTotal
        updateNotification(forceShow, getString(R.string.build_ongoing), desc)
    }

    override fun setTaskProgress(progress: Float) {
        observer?.get()?.setTaskProgress(progress)
    }

    override fun taskSucceed() {
        synchronized(composeResult) {
            composeResult.finished = true
            composeResult.ret = true
            composeResult.failMessage = null
        }
        updateNotification(true, getString(R.string.build_finished), getString(R.string.succeed))
        observer?.get()?.taskSucceed()
    }

    override fun taskFailed(errMessage: String) {
        synchronized(composeResult) {
            composeResult.finished = true
            composeResult.ret = false
            composeResult.failMessage = errMessage
        }
        updateNotification(true, getString(R.string.build_finished), getString(R.string.failed))
        observer?.get()?.taskFailed(errMessage)
    }

    override fun taskWarning(message: String) {
        observer?.get()?.taskWarning(message)
    }

    class ComposeResult {
        var finished = false
        var ret = false
        var failMessage: String? = null
        var curStep: TaskStepInfo? = null
        fun clear() {
            finished = false
            failMessage = null
            curStep = null
        }
    }

    @SuppressLint("HandlerLeak")
    private inner class MyHandler : Handler(Looper.myLooper() ?: Looper.getMainLooper()) {
        private var title: String? = null
        private var desc: String? = null
        fun setInfo(title: String?, desc: String?) {
            this.title = title
            this.desc = desc
        }

        override fun handleMessage(msg: Message) {
            when (msg.what) {
                0 -> {
                    mNotifyBuilder?.let { builder ->
                        title?.let {
                            builder.setContentTitle(it)
                        }
                        desc?.let {
                            builder.setContentText(it)
                        }
                        if (foregroundStarted) {
                            mNotificationManager?.notify(
                                Constants.NOTIFICATION_ID.FOREGROUND_SERVICE,
                                builder.build()
                            )
                        } else {
                            startForeground(
                                Constants.NOTIFICATION_ID.FOREGROUND_SERVICE,
                                builder.build()
                            )
                            foregroundStarted = true
                        }
                    }
                    lastUpdateTime = System.currentTimeMillis()
                }
            }
        }
    }

    inner class ComposeServiceBinder : Binder() {
        fun setObserver(_observer: ITaskCallback) {
            observer = WeakReference(_observer)
            synchronized(composeResult) {
                if (composeResult.finished) {
                    if (composeResult.ret) {
                        _observer.taskSucceed()
                    } else {
                        _observer.taskFailed(composeResult.failMessage)
                    }
                } else {
                    if (composeResult.curStep != null) {
                        _observer.setTaskStepInfo(composeResult.curStep)
                    }
                }
            }
        }

        fun buildAgain() {
            resetStatus()
            showNotification()
            startComposeThread()
        }

        // Stop the build thread
        fun stopBuilding() {
            composeThread?.let { thread ->
                if (thread.isActive) {
                    thread.stopRunning()
                }
            }
            hideNotification()
        }

        fun setBuildHooker(extraMaker: IApkMaking?) {
            this@ApkComposeService.extraMaker = extraMaker
        }

        // Get key/value maps
        val values: Map<String, Any?>
            get() {
                val ret: MutableMap<String, Any?> = HashMap()
                ret["srcApkPath"] = srcApkPath
                ret["targetApkPath"] = targetApkPath
                ret["decodeRootPath"] = decodeRootPath
                modifiedSmaliFolders?.let { codeModified ->
                    ret["codeModified"] = codeModified.isNotEmpty()
                }
                ret["signAPK"] = signAPK
                return ret
            }

        fun hideNotification() {
            mNotificationManager?.let { manager ->
                manager.cancel(Constants.NOTIFICATION_ID.FOREGROUND_SERVICE)
                if (foregroundStarted) {
                    stopForeground(STOP_FOREGROUND_REMOVE)
                    foregroundStarted = false
                }
                composeThread?.let { thread ->
                    if (thread.isActive) {
                        thread.stopRunning()
                    }
                }
                mNotificationManager = null
                Log.e("DEBUG", "notification hidden.")
            }
        }
        fun showNotification() {
            this@ApkComposeService.showNotification()
        }

        // Build thread is still running
        val isRunning: Boolean
            get() = composeThread != null && composeThread!!.isActive
    }
}